File: website/components/symmetry-aware-einsum-contractions/engine/wreath.js
// website/components/symmetry-aware-einsum-contractions/engine/wreath.js
//
// Direct enumerator for the wreath product ∏_i (H_i ≀ S_{m_i}).
//
// `i` indexes identical-operand groups (operands sharing the same name).
// `H_i` is each operand's declared axis symmetry group on its own axes.
// `m_i` is the number of copies of operand i.
//
// Emission order (maintained contract): top-major, base-minor lex. For
// each identical-group, we iterate S_{m_i} in lex order over its
// permutation arrays; for each outer element we iterate the base tuple
// (h_0, …, h_{m-1}) in lex order over `H_i`'s element array. Across
// identical-groups we take the cartesian product.

import { Permutation, dimino } from './permutation.js';
import { parseCycleNotation } from './cycleParser.js';

// Enumerate every element of a rank-n permutation group described by
// `sym` ('symmetric' | 'cyclic' | 'dihedral' | { type: 'custom', generators, axes }
//  | null/'none'/undefined). Returns an array of Permutation on [0, rank).
//
// `sym.generators` for the custom branch may be either:
//   - a raw cycle-notation string ("(0 1), (2 3)"), or
//   - the pre-parsed array-of-cycle-arrays shape that
//     `pipeline.js::analyzeExample` produces via `parseCycleNotation`.
// Both shapes are accepted so callers don't need to re-normalise.
export function enumerateH(sym, rank) {
  if (!sym || sym === 'none') return [Permutation.identity(rank)];

  const identity = Permutation.identity(rank);

  if (sym === 'symmetric' || sym?.type === 'symmetric') {
    const axes = sym?.axes || Array.from({ length: rank }, (_, i) => i);
    // Generators: adjacent transpositions on the axes
    const gens = [];
    for (let k = 0; k < axes.length - 1; k += 1) {
      const arr = Array.from({ length: rank }, (_, i) => i);
      arr[axes[k]] = axes[k + 1];
      arr[axes[k + 1]] = axes[k];
      gens.push(new Permutation(arr));
    }
    return gens.length === 0 ? [identity] : dimino(gens);
  }

  if (sym?.type === 'cyclic' || sym === 'cyclic') {
    const axes = sym?.axes || Array.from({ length: rank }, (_, i) => i);
    if (axes.length <= 1) return [identity];
    const arr = Array.from({ length: rank }, (_, i) => i);
    for (let k = 0; k < axes.length; k += 1) arr[axes[k]] = axes[(k + 1) % axes.length];
    return dimino([new Permutation(arr)]);
  }

  if (sym?.type === 'dihedral' || sym === 'dihedral') {
    const axes = sym?.axes || Array.from({ length: rank }, (_, i) => i);
    if (axes.length <= 2) return [identity];
    // Rotation generator + reflection generator
    const rot = Array.from({ length: rank }, (_, i) => i);
    for (let k = 0; k < axes.length; k += 1) rot[axes[k]] = axes[(k + 1) % axes.length];
    const ref = Array.from({ length: rank }, (_, i) => i);
    for (let k = 0; k < Math.floor(axes.length / 2); k += 1) {
      const a = axes[k];
      const b = axes[axes.length - 1 - k];
      ref[a] = b;
      ref[b] = a;
    }
    return dimino([new Permutation(rot), new Permutation(ref)]);
  }

  if (sym?.type === 'custom') {
    const axes = Array.isArray(sym.axes)
      ? sym.axes
      : Array.from({ length: rank }, (_, i) => i);

    if (axes.length === 0) return [identity];

    let generatorCycles;
    if (typeof sym.generators === 'string') {
      const parsed = parseCycleNotation(sym.generators);
      generatorCycles = parsed.generators;
    } else if (Array.isArray(sym.generators)) {
      generatorCycles = sym.generators;
    } else {
      generatorCycles = null;
    }
    if (!generatorCycles || generatorCycles.length === 0) return [identity];
    const gens = generatorCycles.map((perm) => {
      const arr = Array.from({ length: rank }, (_, i) => i);
      for (const cycle of perm) {
        for (let k = 0; k < cycle.length; k += 1) {
          const fromLocal = cycle[k];
          const toLocal = cycle[(k + 1) % cycle.length];
          const fromAxis = axes[fromLocal];
          const toAxis = axes[toLocal];
          arr[fromAxis] = toAxis;
        }
      }
      return new Permutation(arr);
    });

    return dimino(gens);
  }

  return [identity];
}

// Yield every permutation of [0, m) in lex order.
function* enumerateSymmetric(m) {
  const arr = Array.from({ length: m }, (_, i) => i);
  // Lex-order permutation iteration (Heap's alternative)
  function* permute(k) {
    if (k === m - 1) {
      yield arr.slice();
      return;
    }
    for (let i = k; i < m; i += 1) {
      [arr[k], arr[i]] = [arr[i], arr[k]];
      yield* permute(k + 1);
      [arr[k], arr[i]] = [arr[i], arr[k]];
    }
  }
  if (m === 0) { yield []; return; }
  yield* permute(0);
}

// Given a list of arrays, yield the cartesian product as arrays.
function* cartesianProduct(lists) {
  if (lists.length === 0) { yield []; return; }
  const [head, ...rest] = lists;
  for (const item of head) {
    for (const tail of cartesianProduct(rest)) {
      yield [item, ...tail];
    }
  }
}

// Flatten a wreath element `(baseTuple, topPerm)` for a single
// identical-group into a row-perm on U-vertex indices. `uOffsets[p]`
// is the starting U-vertex index for operand position p. `axisRanks[p]`
// is the rank of operand p.
//
// topPerm[j] = new position of copy j (so if topPerm[0]=1, copy 0 moves
// to position 1). baseTuple[j] is the axis-permutation h_j applied to
// copy j's axes BEFORE the top permutation maps j to its new slot.
//
// Net: U-vertex (operand_position p, axis a) moves to
//   (operand_position topPerm[positionOf(p)], axis baseTuple[j].arr[a]).
function flattenFactorToRowPerm(group, baseTuple, topPerm, uOffsets, axisRanks, nU) {
  const arr = Array.from({ length: nU }, (_, i) => i);
  for (let j = 0; j < group.length; j += 1) {
    const p = group[j];
    const rank = axisRanks[p];
    const newJ = topPerm[j];
    const newP = group[newJ];
    const h = baseTuple[j];
    for (let a = 0; a < rank; a += 1) {
      const from = uOffsets[p] + a;
      const to = uOffsets[newP] + h.arr[a];
      arr[to] = from;
    }
  }
  return arr;
}

// Compose row-perms from per-group factors. Each factor has already
// been written into a full-length array; compose by arr[i] = accumulated
// where we apply factors left-to-right. Since different identical-groups
// touch disjoint U-vertex ranges, composition is a single merge.
function mergeFactors(factorArrs, nU) {
  const out = Array.from({ length: nU }, (_, i) => i);
  for (const arr of factorArrs) {
    for (let i = 0; i < nU; i += 1) {
      if (arr[i] !== i) out[i] = arr[i];
    }
  }
  return out;
}

/**
 * Main entry point.
 *
 * @param {object} args
 * @param {number[][]} args.identicalGroups - array of groups; each group is
 *   an array of operand positions (0-indexed) that share the same operand
 *   name.
 * @param {Array<string|object|null>} args.perOpSymmetry - per-operand
 *   declared axis symmetry (same shape as pipeline.js produces).
 * @param {number[]} args.axisRanks - rank of each operand position.
 * @param {number[]=} args.uOffsets - starting U-vertex index for each
 *   operand. If omitted, computed as the running sum of axisRanks.
 *
 * @yields {{ rowPerm: Permutation, factorization: Array<{baseTuple: Permutation[], topPerm: number[]}> }}
 *   One object per wreath element, in top-major / base-minor lex order.
 */
export function* enumerateWreath({ identicalGroups, perOpSymmetry, axisRanks, uOffsets }) {
  const nOperands = axisRanks.length;
  const offsets = uOffsets || (() => {
    const o = new Array(nOperands);
    let acc = 0;
    for (let p = 0; p < nOperands; p += 1) { o[p] = acc; acc += axisRanks[p]; }
    return o;
  })();
  const nU = offsets[nOperands - 1] + axisRanks[nOperands - 1];

  // For each identical-group, build a per-group factor enumerator.
  const factorIterables = identicalGroups.map((group) => {
    const m = group.length;
    const rank = axisRanks[group[0]];
    const sym = perOpSymmetry[group[0]];
    const H = enumerateH(sym, rank);

    // Per-group wreath order = |H|^m · m!
    return (function* () {
      for (const topPerm of enumerateSymmetric(m)) {
        const baseTupleLists = new Array(m).fill(H);
        for (const baseTuple of cartesianProduct(baseTupleLists)) {
          const arr = flattenFactorToRowPerm(group, baseTuple, topPerm, offsets, axisRanks, nU);
          yield { arr, baseTuple, topPerm: topPerm.slice() };
        }
      }
    })();
  });

  // Cartesian across identical-groups. For small counts this is fine.
  // (In practice nGroups is ≤ 3 on all 22 presets.)
  const factorArrays = factorIterables.map((iter) => [...iter]);
  for (const combo of cartesianProduct(factorArrays)) {
    const arrs = combo.map((f) => f.arr);
    const merged = mergeFactors(arrs, nU);
    yield {
      rowPerm: new Permutation(merged),
      factorization: combo.map((f) => ({ baseTuple: f.baseTuple, topPerm: f.topPerm })),
    };
  }
}

File: src/flopscope/_accumulation/_wreath.py
"""Direct enumerator for the wreath product ∏_i (H_i ≀ S_{m_i}).

Port of website/components/symmetry-aware-einsum-contractions/engine/wreath.js.

`i` indexes identical-operand groups (operands sharing the same name).
`H_i` is each operand's declared axis symmetry on its own axes.
`m_i` is the number of copies of operand i.

Each wreath element is a row permutation σ on the U-vertices, paired with
factorization metadata that names the (outer-S_{m_i}, base-H_i) decomposition
for diagnostic display.
"""

from __future__ import annotations

import itertools
import math
from collections.abc import Iterator, Sequence
from dataclasses import dataclass
from typing import Any

from flopscope._perm_group import SymmetryGroup
from flopscope._perm_group import _dimino
from flopscope._perm_group import _Permutation as Permutation


@dataclass(frozen=True)
class WreathElement:
    """One element of ∏_i (H_i ≀ S_{m_i}). Carries the row permutation and provenance."""
    row_perm: Permutation
    factorization: dict[str, Any]


def enumerate_h(sym: Any, rank: int) -> Iterator[Permutation]:
    """Enumerate every element of the rank-`rank` group described by `sym`.

    `sym` may be:
      - None or 'none' → identity only
      - 'symmetric' → S_rank
      - 'cyclic' → C_rank
      - 'dihedral' → D_rank
      - a SymmetryGroup object → use its elements directly
      - a dict with {'type': 'custom', 'generators': [...]} → custom group

    Returns rank-`rank` permutations (operate on positions 0..rank-1).
    """
    identity = Permutation.identity(rank)
    if sym is None or sym == 'none':
        yield identity
        return

    if isinstance(sym, SymmetryGroup):
        axes = getattr(sym, 'axes', None)
        for el in sym.elements():
            arr = list(range(rank))
            if axes is not None:
                # elements() yields permutations on positions 0..len(axes)-1;
                # map them to the actual tensor axis positions via sym.axes.
                for local_i, local_j in enumerate(el.array_form):
                    if local_i < len(axes) and local_j < len(axes):
                        from_axis = axes[local_i]
                        to_axis = axes[local_j]
                        if from_axis < rank and to_axis < rank:
                            arr[from_axis] = to_axis
            else:
                # No axes annotation; embed at zero offset (legacy path).
                for i, j in enumerate(el.array_form):
                    if i < rank:
                        arr[i] = j if j < rank else i
            yield Permutation(arr)
        return

    if sym == 'symmetric' or (isinstance(sym, dict) and sym.get('type') == 'symmetric'):
        gens = []
        for k in range(rank - 1):
            arr = list(range(rank))
            arr[k], arr[k + 1] = arr[k + 1], arr[k]
            gens.append(Permutation(arr))
        if not gens:
            yield identity
            return
        for el in _dimino(tuple(gens)):
            yield el
        return

    if sym == 'cyclic' or (isinstance(sym, dict) and sym.get('type') == 'cyclic'):
        if rank <= 1:
            yield identity
            return
        rotation = list(range(1, rank)) + [0]
        for el in _dimino((Permutation(rotation),)):
            yield el
        return

    if sym == 'dihedral' or (isinstance(sym, dict) and sym.get('type') == 'dihedral'):
        if rank <= 2:
            for el in enumerate_h('symmetric', rank):
                yield el
            return
        rot = list(range(1, rank)) + [0]
        ref = list(range(rank))
        for k in range(rank // 2):
            ref[k], ref[rank - 1 - k] = ref[rank - 1 - k], ref[k]
        for el in _dimino((Permutation(rot), Permutation(ref))):
            yield el
        return

    raise ValueError(f'unsupported symmetry declaration: {sym!r}')


def _outer_permutations(m: int) -> Iterator[list[int]]:
    """Yield every permutation array of length m (S_m). Used for the outer factor."""
    for perm_tuple in itertools.permutations(range(m)):
        yield list(perm_tuple)


def _flatten_factor_to_row_perm(
    group: Sequence[int],
    base_tuple: Sequence[Permutation],
    top_perm: Sequence[int],
    u_offsets: Sequence[int],
    axis_ranks: Sequence[int],
    n_u: int,
) -> list[int]:
    """Build row-perm contribution for one identical-group factor.

    Mirrors JS flattenFactorToRowPerm: arr[to] = from (inverse representation
    consistent with the JS engine).
      top_perm[j] = new position of copy j within the group.
      base_tuple[j] = axis permutation applied to copy j's axes before relocation.
    """
    arr = list(range(n_u))
    for j in range(len(group)):
        p = group[j]
        rank = axis_ranks[p]
        new_j = top_perm[j]
        new_p = group[new_j]
        h = base_tuple[j]
        for a in range(rank):
            from_idx = u_offsets[p] + a
            to_idx = u_offsets[new_p] + h.array_form[a]
            arr[to_idx] = from_idx
    return arr


def enumerate_wreath(
    *,
    identical_groups: Sequence[Sequence[int]],
    per_op_symmetry: Sequence[Any],
    axis_ranks: Sequence[int],
    u_offsets: Sequence[int],
) -> Iterator[WreathElement]:
    """Iterate ∏_i (H_i ≀ S_{m_i}) and yield row permutations on the U-vertices.

    `identical_groups`: tuple of operand-index tuples, each grouping copies of
        the same operand.
    `per_op_symmetry`: parallel to operand index — declared H_i for each operand.
    `axis_ranks`: parallel — number of axes per operand.
    `u_offsets`: parallel — starting U-vertex index for each operand.
    """
    total_u = sum(axis_ranks)

    # For each identical-group, build a list of (arr, factor_meta) pairs.
    per_group_options: list[list[tuple[list[int], dict]]] = []
    for grp in identical_groups:
        m = len(grp)
        # Base H_i — all copies in the group share the same declared symmetry.
        base_sym = per_op_symmetry[grp[0]]
        base_rank = axis_ranks[grp[0]]
        h_elements = list(enumerate_h(base_sym, base_rank))

        group_options: list[tuple[list[int], dict]] = []
        for top_perm in _outer_permutations(m):
            for base_tuple in itertools.product(h_elements, repeat=m):
                arr = _flatten_factor_to_row_perm(
                    grp, base_tuple, top_perm, u_offsets, axis_ranks, total_u
                )
                group_options.append((
                    arr,
                    {
                        'group': tuple(grp),
                        'outer': tuple(top_perm),
                        'base_arrs': tuple(tuple(h.array_form) for h in base_tuple),
                    },
                ))
        per_group_options.append(group_options)

    if not per_group_options:
        # No operands: just identity.
        yield WreathElement(
            row_perm=Permutation.identity(total_u),
            factorization={'groups': ()},
        )
        return

    # Cartesian product across groups: merge contributions into a single row perm.
    # Different groups touch disjoint U-vertex ranges, so merge = overwrite non-identity.
    for combo in itertools.product(*per_group_options):
        row_perm_arr = list(range(total_u))
        for arr, _factor in combo:
            for i in range(total_u):
                if arr[i] != i:
                    row_perm_arr[i] = arr[i]
        factorization = {
            'groups': tuple(factor for _, factor in combo),
        }
        yield WreathElement(
            row_perm=Permutation(row_perm_arr),
            factorization=factorization,
        )
